In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'


engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
house_prices_df = pd.read_sql_query('select * from houseprices',con=engine)

engine.dispose()

In [2]:
house_prices_df = pd.concat([house_prices_df,pd.get_dummies(house_prices_df.mszoning, prefix="mszoning", drop_first=True)], axis=1)
house_prices_df = pd.concat([house_prices_df,pd.get_dummies(house_prices_df.street, prefix="street", drop_first=True)], axis=1)
dummy_column_names = list(pd.get_dummies(house_prices_df.mszoning, prefix="mszoning", drop_first=True).columns)
dummy_column_names = dummy_column_names + list(pd.get_dummies(house_prices_df.street, prefix="street", drop_first=True).columns)

In [3]:
X = house_prices_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf'] + dummy_column_names]
Y = house_prices_df.saleprice

In [4]:
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.767
Method:                 Least Squares   F-statistic:                     482.0
Date:                Sun, 18 Aug 2019   Prob (F-statistic):               0.00
Time:                        00:57:05   Log-Likelihood:                -17475.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1449   BIC:                         3.503e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.173e+05    1.8e+04     -6.502      0.000   -1.53e+05   -8.19e+04
overallqual  2.333e+04   1088.506     21.430      0.000    2.12e+04    2.55e+04
grlivarea      45.6344      2.468     18.494      0.000      40.794      50.475
garagecars   1.345e+04   2990.453      4.498      0.000    7584.056    1.93e+04
garagearea     16.4082     10.402      1.577      0.115      -3.997      36.813
totalbsmtsf    28.3816      2.931      9.684      0.000      22.633      34.131
mszoning_FV  2.509e+04   1.37e+04      1.833      0.067   -1761.679    5.19e+04
mszoning_RH  1.342e+04   1.58e+04      0.847      0.397   -1.77e+04    4.45e+04
mszoning_RL  2.857e+04   1.27e+04      2.246      0.025    3612.782    5.35e+04
mszoning_RM  9047.7894   1.29e+04      0.702      0.483   -1.62e+04    3.43e+04
street_Pave   151.0376   1.63e+04      0.009      0.993   -3.18e+04    3.21e+04
==============================================================================
Omnibus:                      415.883   Durbin-Watson:                   1.979
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            41281.526
Skew:                          -0.115   Prob(JB):                         0.00
Kurtosis:                      29.049   Cond. No.                     5.53e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.53e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The model has a very low bias very close to zero.

- An increase of the overall material and finish quality of the house by 1 unit results in increasing its price by 23,330 Dollars

- An increase of the ground living area of the house by 1 square feet results in increasing its price by 45.63 Dollars

- An increase of the garage size by 1 car results in increasing its price by 13,450 Dollars

- An increase of the garage size by 1 square feet results in increasing its price by 16.41 Dollars

- An increase of the total basement size by 1 square feet results in increasing its price by 28.38 Dollars

- If the zoning classification of the house is Floating Village Residential (FV), the price of the house increases by 25,090 Dollars

- If the zoning classification of the house is Residential High Density (RH), the price of the house increases by 13,420 Dollars

- If the zoning classification of the house is Residential Low Density (RL), the price of the house increases by 28,570 Dollars

- If the zoning classification of the house is Residential Medium Density (RM), the price of the house increases by 9,047.79 Dollars

- If the access road to the house is paved, the price of the house increases by 151.04 Dollars


The statistically significant features are:
    
- overallqual , grlivarea, garagecars, totalbsmfsf, mszoning_RL

The coeficients which are insignificant are:
    
- garagearea, mszoning_FV, mszoning_RH, mszoning_RM, street_Pave

In [7]:
X2 = house_prices_df[['overallqual', 'grlivarea', 'garagecars', 'totalbsmtsf', 'mszoning_RL']]

In [8]:
X2 = sm.add_constant(X2)

results = sm.OLS(Y, X2).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.767
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     956.8
Date:                Sun, 18 Aug 2019   Prob (F-statistic):               0.00
Time:                        01:23:18   Log-Likelihood:                -17481.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1454   BIC:                         3.501e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.083e+05   4804.236    -22.540      0.000   -1.18e+05   -9.89e+04
overallqual  2.396e+04   1060.549     22.588      0.000    2.19e+04     2.6e+04
grlivarea      45.4093      2.452     18.517      0.000      40.599      50.220
garagecars   1.763e+04   1731.766     10.183      0.000    1.42e+04     2.1e+04
totalbsmtsf    28.8729      2.862     10.088      0.000      23.259      34.487
mszoning_RL  1.596e+04   2558.589      6.238      0.000    1.09e+04     2.1e+04
==============================================================================
Omnibus:                      402.656   Durbin-Watson:                   1.979
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            35429.680
Skew:                          -0.080   Prob(JB):                         0.00
Kurtosis:                      27.133   Cond. No.                     9.53e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.53e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

No big changes interms of coefficients except the coefficnets of mszoning_RL which dropped from 28,570 to 15,960 (almost 50% decrease), and the coefficient of garagecars which increased from 13,450 to 17,630 (almost 30% increase)

The model has a very low bias very close to zero.

- An increase of the overall material and finish quality of the house by 1 unit results in increasing its price by 23,960 Dollars

- An increase of the ground living area of the house by 1 square feet results in increasing its price by 45.41 Dollars

- An increase of the garage size by 1 car results in increasing its price by 17,630 Dollars

- An increase of the total basement size by 1 square feet results in increasing its price by 28.87 Dollars

- If the zoning classification of the house is Residential Low Density (RL), the price of the house increases by 15,960 Dollars

From the coefficients we have in the model, the overallqual feature lloks like the prominent factor but this depends on the range of increase or decrease for each variable. For example, if the current overallqual score is 9, we do not have area of improvement other than 1 which will result in increasing the house price by 23,960 Dollars, but if for example we increase the garage size by 2 cars, the price of the house increases by 35,260 Dollars and in this case the garagecars will be more prominent factor